README.md
TAPAS PhysIO Toolbox Version 2017
=================================

> Copyright (C) 2012-2017 Lars Kasper <kasper@biomed.ee.ethz.ch>

> Translational Neuromodeling Unit (TNU)

> Institute for Biomedical Engineering

> University of Zurich and ETH Zurich

Copying
-------

The PhysIO Toolbox is free software: you can redistribute it and/or
modify it under the terms of the GNU General Public License as
published by the Free Software Foundation, either version 3 of the
License, or (at your option) any later version.

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program (see the file COPYING).  If not, see
<http://www.gnu.org/licenses/>.

Purpose
-------

The PhysIO Toolbox provides model-based physiological noise correction of 
fMRI data using peripheral measures of respiration and cardiac pulsation. 
It incorporates noise models of cardiac/respiratory phase (RETROICOR, 
Glover et al. 2000), as well as heart rate variability and respiratory 
volume per time (cardiac response function, Chang et. al, 2009, respiratory 
response function, Birn et al. 2006). The toolbox is usable via the SPM 
batch editor, performs automatic pre-processing of noisy peripheral data 
and outputs nuisance regressor files directly suitable for SPM 
(multiple_regressors.txt).

Installation
------------

### Matlab ###
- Unzip the TAPAS archive
- Add tapas/physio/code to your matlab path

### SPM ###
- Certain functionality (Batch Editor GUI, pipeline dependencies, model assessment via F-contrasts) require the installation of SPM
- Afterwards, the PhysIO Toolbox has to be registered as an SPM toolbox by copying the `physio/code` folder to `spm/toolbox/physio`


Getting Started
---------------

Run `main_ECG3T.m` in subdirectory `physio/examples` of the toolbox
See subdirectory `physio/docs` and next section of this document.


Getting Help/Documentation
--------------------------

Several documentation files are provided with this toolbox. They have the extension .md (markdown), i.e. are plain text files, but can be conveniently viewed online as the github/gitlab Wiki.
You can find them in `physio/docs`.

List of Documentation files:
- README.md: this file, purpose, installation, getting started, pointer to more help
- FAQ.md: Frequently asked questions (for users)
- QUICKSTART.md: Example script and how to use on test data, Intro to Batch Editor GUI
- MANUAL.md: Reference Manual (mostly for developers) listing all functions, and rationales of the toolbox, disecting its modular structure
- WIKIMAIN.md: Landing Page of Wiki. Navigation to all other files and this explanation
- EXAMPLES.md: List and explanation of all examples
- CHANGELOG.md: List of all toolbox versions and the respective release notes, i.e. major changes in functionality, bugfixes etc.


Background
----------

The PhysIO Toolbox provides physiological noise correction for fMRI-data 
from peripheral measures (ECG/pulse oximetry, breathing belt). It is 
model-based, i.e. creates nuisance regressors from the physiological 
monitoring that can enter a General Linear Model (GLM) analysis, e.g. 
SPM8/12. Furthermore, for scanner vendor logfiles (PHILIPS, GE, Siemens), 
it provides means to statistically assess peripheral data (e.g. heart rate variability) 
and recover imperfect measures (e.g. distorted R-peaks of the ECG).

Facts about physiological noise in fMRI:
- Physiological noise can explain 20-60 % of variance in fMRI voxel time 
  series (Birn2006, Hutton2011, Harvey2008.
    - Physiological noise affects a lot of brain regions (s. figure, e.g. 
      brainstem or OFC), especially next to CSF, arteries (Hutton2011). 
    - If not accounted for, this is a key factor limiting sensitivity for effects of interest.
- Physiological noise contributions increase with field strength; they 
  become a particular concern at and above 3 Tesla (Kasper2009, Hutton2011).
- In resting state fMRI, disregarding physiological noise leads to wrong 
  connectivity results (Birn2006).

Therefore, some kind of physiological noise correction is highly recommended for every statistical fMRI analysis.

Model-based correction of physiological noise: 
- Physiological noise can be decomposed into periodic time series following 
  heart rate and breathing cycle.
- The Fourier expansion of cardiac and respiratory phases was introduced as 
  RETROICOR (RETROspective Image CORrection, Glover2000, 
  see also Josephs1997).
- These Fourier Terms can enter a General Linear Model (GLM) as nuisance 
  regressors, analogous to movement parameters.
- As the physiological noise regressors augment the GLM and explain 
  variance in the time series, they increase sensitivity in all contrasts 
  of interest.

		
Features of this Toolbox
------------------------

### Physiological Noise Modeling ###

- Modeling physiological noise regressors from peripheral data 
  (breathing belt, ECG, pulse oximeter) 
- State of the art RETROICOR cardiac and respiratory phase expansion
- Cardiac response function (Chang et al, 2009) and respiratory response 
  function (Birn et al. 2006) modelling of heart-rate variability and 
  respiratory volume  per time influence on physiological noise
- Flexible expansion orders to model different contributions of cardiac, 
  respiratory and interaction terms (see Harvey2008, Hutton2011)
- Automatic creation of nuisance regressors, full integration into standard 
  GLMs, tested for SPM8/12 ("multiple_regressors.mat")

### Flexible Read-in ###

The toolbox is dedicated to seamless integration into a clinical research s
etting and therefore offers correction methods to recover physiological 
data from imperfect peripheral measures.
- General Electric logfiles
- Siemens logfiles (synchronization via DICOM time stamp or tics time scale)
- Philips SCANPHYSLOG-file handling, including automatic alignment of scan volume timing and physiological time series 
  from logged gradient timecourses
- BioPac .mat-files


Compatibility and Support
-------------------------

- Matlab Toolbox
- Input: 
    - Fully integrated to work with physiological logfiles for Philips MR systems (SCANPHYSLOG)
    - tested for General Electric (GE) log-files
    - implementation for Siemens log-files
    - also: interface for 'Custom', i.e. general heart-beat time stamps 
      & breathing volume time courses from other log formats
- Output: 
    - Nuisance regressors for mass-univariate statistical analysis with SPM5,8,12
      or as text file for export to any other package
    - raw and processed physiological logfile data
- Part of the TNU Software Edition: long term support and ongoing development


Contributors
------------

- Lead Programmer: Lars Kasper, TNU & MR-Technology Group, IBT, University & ETH Zurich
- Project Team: Steffen Bollmann, Children's Hospital Zurich & ETH Zurich
- Contributors:
    - Jakob Heinzle, TNU Zurich
    - Eduardo Aponte, TNU Zurich


Contact
-------
Send bug reports and suggestions to our mailing list: tapas@sympa.ethz.ch


References
----------

1. Glover, G.H., Li, T.Q. & Ress, D. Image-­‐based method for retrospective correction
of PhysIOlogical motion effects in fMRI: RETROICOR. Magn Reson Med 44, 162-­‐
7 (2000).

2. Hutton, C. et al. The impact of PhysIOlogical noise correction on fMRI at 7 T.
NeuroImage 57, 101-­‐112 (2011).

3. Harvey, A.K. et al. Brainstem functional magnetic resonance imaging:
Disentangling signal from PhysIOlogical noise. Journal of Magnetic Resonance
Imaging 28, 1337-­‐1344 (2008).

4. Kasper, L., Bollmann, S., Diaconescu, A.O., Hutton, C., Heinzle, J., Iglesias, S.,
Hauser, T.U., Sebold, M., Manjaly, Z.-M., Pruessmann, K.P., Stephan, K.E., 2016.
The PhysIO Toolbox for Modeling PhysIOlogical Noise in fMRI Data. Journal of
Neuroscience Methods accepted. doi:10.1016/j.jneumeth.2016.10.019

5. Behzadi, Y., Restom, K., Liau, J., Liu, T.T., 2007. A component based noise
correction method (CompCor) for BOLD and perfusion based fMRI. NeuroImage 37,
90–101. doi:10.1016/j.neuroimage.2007.04.042
    
6. Birn, R.M., Smith, M.A., Jones, T.B., Bandettini, P.A., 2008. The respiration response
function: The temporal dynamics of fMRI s ignal fluctuations related to changes in
respiration. NeuroImage 40, 644–654. doi:10.1016/j.neuroimage.2007.11.059
    
7. Chang, C., Cunningham, J.P., Glover, G.H., 2009. Influence of heart rate on the
BOLD signal: The cardiac response function. NeuroImage 44, 857–869.
doi:10.1016/j.neuroimage.2008.09.029
    
8. Siegel, J.S., Power, J.D., Dubis, J.W., Vogel, A.C., Church, J.A., Schlaggar, B.L.,
Petersen, S.E., 2014. Statistical improvements in functional magnetic resonance
imaging analyses produced by censoring high-motion data points. Hum. Brain Mapp.
35, 1981–1996. doi:10.1002/hbm.22307

CHANGELOG
DOWNLOADS & RELEASE INFORMATION
===============================

Current Release
---------------

PhysIO_Toolbox_R2017.1

February 19, 2017

Minor Release Notes (R2017.1)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

* Substantially improved Siemens interface, both for VB/VD and 3T/7T releases
** several bugfixes
** based on extensive user feedback from Berlin and Brisbane
* New functionality tapas_physio_overlay_contrasts.m to display non-physio 
  contrasts automatically as well

Major Release Notes (r904 / R2016.1):
- Software version for accepted PhysIO Toolbox Paper: doi:10.1016/j.jneumeth.2016.10.019
- Tested and expanded versions of examples
- Improved stability by bugfixes and compatibility to Matlab R2016
- Slice-wise regressor creation
- Detection of constant physiological time series (detachment, clipping)
- Refactoring of report_contrasts and compute_tsnr_gains as standalone functionality
- Improved Read-in capabilities (Siemens respiration data, BioPac .mat)
- Migration from svn (r904) to git (tnurepository) for version control

Major Release Notes (r835):
- Software version for Toolbox Paper submission
- Noise ROIs modeling
- Extended motion models (24 parameters, Volterra expansion)
- HRV/RVT models with optional multiple delay regressors
- Report_contrasts with automatic contrast generation for all regressor groups
- compute_tsnr_gains for individual physiological regressor groups
- consistent module naming (scan_timing, preproc)
- Visualisation improvement (color schemes, legends)

Minor Release Notes (r666):
- Compatibility tested for SPM12, small bugfixes Batch Dependencies
- Cleaner Batch Interface with grouped sub-menus (cfg_choice)
- new model: 'none' to just read out physiological raw data and preprocess,
  without noise modelling 
- Philips: Scan-timing via gradient log now automatized (gradient_log_auto)
- Siemens: Tics-Logfile read-in (proprietary, needs Siemens-agreement)
- All peak detections (cardiac/respiratory) now via auto_matched algorithm
- Adapt plots/saving for Matlab R2014b

Major Release Notes (r534):
- Read-in of Siemens plain text log files; new example dataset for Siemens
- Speed up and debugging of auto-detection method for noisy cardiac data => new method thresh.cardiac.initial_cpulse_select.method = ???auto_matched???
- Error handling for temporary breathing belt failures (Eduardo Aponte, TNU Zurich)
- slice-wise regressors can be created by setting sqpar.onset_slice to a index vector of slices

Major Release Notes (r497):
- SPM matlabbatch GUI implemented (Call via Batch -> SPM -> Tools -> TAPAS PhysIO Toolbox)
- improved, automatic heartbeat detection for noisy ECG now standard for ECG and Pulse oximetry (courtesy of Steffen Bollmann)
- QuickStart-Manual and PhysIO-Background presentation expanded/updated
- job .m/.mat-files created for all example datasets
- bugfixes cpulse-initial-select method-handling (auto/manual/load)

Major Release Notes (r429):
- Cardiac and Respiratory response function regressors integrated in workflow (heart rate and breathing volume computation)
- Handling of Cardiac and Respiratory Logfiles only
- expanded documentation (Quickstart.pdf and Handbook.pdf)
- read-in of custom log files, e.g. for BrainVoyager peripheral data
- more informative plots and commenting (especially in tapas_physio_new).

Minor Release Notes (r354):
- computation of heart and breathing rate in Philips/PPU/main_PPU.m
- prefix of functions with tapas_*

Major Release Notes (r241):
- complete modularization of reading/preprocessing/regressor creation for peripheral physiological data
- manual selection of missed heartbeats in ECG/pulse oximetry (courtesy of Jakob Heinzle)
- support for logfiles from GE scanners (courtesy of Steffen Bollmann, KiSpi Zuerich)
- improved detection of pulse oximetry peaks (courtesy of Steffen Bollmann)
- improved documentation
- consistent function names (prefixed by "physio_")

NOTE: Your main_ECG/PPU.m etc. scripts from previous versions (<=r159) will not work with this one any more. Please adapt one of the example scripts for your needs (~5 min of work). The main benefit of this version is a complete new variable structure that is more sustainable and makes the code more readable.



code/physio_new.m
function physio = physio_new(default_scheme, physio_in)
% creates complete PhysIO structure fed into physio_main_create_regressors
%
%    physio = physio_new(default_scheme, physio_in)
%
% IN
%   default_scheme  - if set, default values for structure entries are set
%                       according to the application
%                       different templates are predefined, e.g.
%                       'empty' (default) - all strings are set to '', all
%                                     numbers to []
%                       'RETROICOR' order of RETROICOR expansion taken from 
%                       Harvey2008, JRMI28(6), p1337ff.
%                       'scan_timing_from_start'
%                       'manual_peak_select'
%   physio_in       - used as input, only fields related to default_scheme
%                     are overwritten, the others are kept as in physio_in
%
% OUT
%   physio          - the complete physio structure, which can be unsed in
%                     physio_main_create_regressors
%
% NOTE
%   All parameters used in the physIO toolbox are defined AND DOCUMENTED in
%   this file. Just scroll down and read through the comments!
% 
% EXAMPLE
%   physio = physio_new('empty')
%   physio = physio_new('RETROICOR');
%   physio = physio_new('manual_peak_select', physio);
%
%   See also physio_main_create_regressors
%
% Author: Lars Kasper
% Created: 2013-04-23
% Copyright (C) 2013 TNU, Institute for Biomedical Engineering, University of Zurich and ETH Zurich.
%
% This file is part of the TNU CheckPhysRETROICOR toolbox, which is released under the terms of the GNU General Public
% Licence (GPL), version 3. You can redistribute it and/or modify it under the terms of the GPL
% (either version 3 or, at your option, any later version). For further details, see the file
% COPYING or <http://www.gnu.org/licenses/>.
%
% $Id$

% if not specified differently, create everything empty
if ~nargin
    default_scheme = 'empty';
end

if nargin >= 2
    log_files   = physio_in.log_files;
    thresh  = physio_in.thresh;
    sqpar   = physio_in.sqpar;
    model   = physio_in.model;
    verbose = physio_in.verbose;
else
    
    
    %% log_files
    % structure containing general physiological log-file information
    log_files.vendor       = ''; % 'Philips', 'GE', or 'Siemens', depending on your
                                 %  MR Scanner system
    log_files.cardiac      = ''; % 'SCANPHYSLOG.log'; logfile with cardiac data
    log_files.respiration  = ''; % 'SCANPHYSLOG.log'; logfile with respiratory data
                                 %                    (same as .cardiac for Philips)
  % log_files.sampling_interval = [];   % in seconds, 2e-3 for Philips, variable for GE,
                                        % e.g. 40e-3
 
                                        
    %% sqpar
    sqpar.Nslices           = [];   % number of slices per volume in fMRI scan
    sqpar.NslicesPerBeat    = [];   % usually equals Nslices, unless you trigger with the heart beat
    sqpar.TR                = [];   % volume repetition time in seconds
    sqpar.Ndummies          = [];   % number of dummy volumes
    sqpar.Nscans            = [];   % number of full volumes saved (volumes in nifti file,
                                    % usually rows in your design matrix)
    sqpar.Nprep             = [];   % set to >=0 to count scans and dummy
                                    % number of non-dummy, volume like preparation pulses
                                    % before 1st dummy scan. If set, logfile is read from beginning,
                                    % otherwise volumes are counted from last detected volume in the logfile
    sqpar.TimeSliceToSlice  = [];   % time between the acquisition of 2 subsequent
                                    % slices; typically TR/Nslices or minTR/Nslices, 
                                    % if minimal temporal slice spacing was chosen
                                    % NOTE: only necessary, if thresh.grad_direction 
                                    % is empty and nominal scan timing is used
    sqpar.onset_slice       = 19;   % slice whose scan onset determines the adjustment of the
                                    % regressor timing to a particular slice for the whole volume
                                    % volumes from beginning of run, i.e. logfile,
                                    % includes counting of preparation gradients
                              
    
    %% thresh
    % determines thresholds used in preprocessing physiological logfiles,
    % either their timing (thresh.scan_timing) or the peripheral measures
    % itself (thresh.cardiac, thresh.respiration)
    thresh.scan_timing = [];    % leave empty, if nominal scan timing, 
                                % derived from sqpar, shall be used

    thresh.scan_timing.grad_direction = ''; % 'x', 'y', or 'z'; 
                                            % if set, sequence timing is calculated 
                                            % from logged gradient timecourse along 
                                            % this coordinate axis;
    thresh.scan_timing.zero     = [];   % gradient values below this value are set to zero;
                                        % should be those which are unrelated to slice acquisition start
    thresh.scan_timing.slice    = [];   % minimum gradient amplitude to be exceeded when a slice scan starts
    thresh.scan_timing.vol      = [];   % minimum gradient amplitude to be exceeded when a new
                                        % volume scan starts; 
                                        % leave [], if volume events shall be determined as
                                        % every Nslices-th scan event or via vol_spacing
    thresh.vol_spacing          = [];   % duration (in seconds) from last slice acq to
                                        % first slice of next volume;
                                        % leave [], if .vol-threshold shall be used
                                        
    thresh.cardiac = [];
    thresh.cardiac.modality = ''; % 'ECG','ECG_raw', or 'OXY' (for pulse oximetry), 'OXY_OLD', [deprecated]
    
    % The initial cardiac pulse selection structure: Determines how the
    % majority of cardiac pulses is detected
    thresh.cardiac.initial_cpulse_select.method = 'load_from_logfile'; % 'load_from_logfile', 'manual' (rather: threshold...autocorrelate?), 'load'
    thresh.cardiac.initial_cpulse_select.file = ''; % file containing reference ECG-peak (variable kRpeak)
                                                    % used for method 'manual' or 'load' [default: not set] string of file containing a
                                                    % if method == 'manual', this file is saved after picking the QRS-wave
                                                    % such that results are reproducible                                    
    thresh.cardiac.initial_cpulse_select.min = [];  % threshold for correlation with QRS-wave to find cardiac pulses 
    thresh.cardiac.initial_cpulse_select.kRpeak = []; % variable saving an example cardiac QRS-wave to correlate with ECG time series
    
    % The posthoc cardiac pulse selection structure: If only few (<20)
    % cardiac pulses are missing in a session due to bad signal quality, a
    % manual selection after visual inspection is possible using the
    % following parameters. The results are saved for reproducibility
    thresh.cardiac.posthoc_cpulse_select.method = 'off'; % 'off', 'manual', 'load'
                                                         % 'off' - no manual selection of peaks
                                                         % 'manual' - pick and save additional peaks manually
                                                         % 'load' - load previously selected cardiac pulses                                                          
    thresh.cardiac.posthoc_cpulse_select.file = '';  % filename where cardiac pulses are saved after manual picking      
    
    % Suspicious positions of missing or too many cardiac pulses are
    % pre-selected by detecting outliers in histogram of
    % heart-beat-2-beat-intervals
    thresh.cardiac.posthoc_cpulse_select.percentile = 80; % percentile of beat-2-beat interval histogram that constitutes the "average heart beat duration" in the session
    thresh.cardiac.posthoc_cpulse_select.upperThresh = 60; % minimum exceedance (in %) from average heartbeat duration to be classified as missing heartbeat
    thresh.cardiac.posthoc_cpulse_select.lowerThresh = 60; % minimum reduction (in %) from average heartbeat duration to be classified an abundant heartbeat
    
    
    %% model
    % Determines the physiological noise model derived from preprocessed physiological data
    model.type = '';                            % 'RETROICOR' - as in Glover el al, MRM 44, 2000
    model.input_other_multiple_regressors = ''; % other nuisance regressors to be included in design matrix
                                                % either txt-file or mat-file with variable R
    model.output_multiple_regressors = '';      % output file for usage in SPM multiple_regressors GLM-specification
                                                % either txt-file or mat-file with variable R
    model.order.c = [];                         % natural number, order of cardiac phase Fourier expansion
    model.order.r = [];                         % natural number, order of respiratory phase Fourier expansion
    model.order.cr = [];                        % natural number, order of cardiac-respiratory-phase-interaction Fourier expansion
                                                % See Harvey et al, JMRI 28, 2008
    model.order.orthogonalise = 'none';         % string indicating which regressors shall be orthogonalised; 
                                                % mainly needed, if acquisition was triggered to heartbeat (set to 'cardiac') OR
                                                % if session mean shall be evaluated (e.g. SFNR-studies, set to 'all')
                                                % 'n' or 'none'     - no orthogonalisation is performed
                                                % Possible Values (default: 'none'
                                                %   'c' or 'cardiac'  - only cardiac regressors are orthogonalised
                                                %   'r' or 'resp'     - only respiration regressors are orthogonalised
                                                %   'mult'            - only multiplicative regressors are orthogonalised
                                                %   'all'             - all physiological regressors are orthogonalised to each other

                                                
    %% verbose
    % determines how many figures shall be generated to follow the workflow
    % of the toolbox and whether the graphical output shall be saved (to a
    % PostScript-file)
    verbose.level = 1;            % 0 = no graphical output; 1 = main plots (default);  
                                  % 2 = debugging plots, for setting up new study; 3 = all plots
    verbose.fig_handles = [];     % collector of all generated figure handles during a run of physio_main_create_regressors
    verbose.fig_output_file = ''; % file name (including extension) where to print all physIO output figures to,
                                  % e.g. 'PhysIO_output.ps' or 'PhysIO_output.jpg'
                                  % The specified extension determines how the
                                  % figures will be saved
                                  %     .ps - all figures are saved to the
                                  %     same, multiple-page postscript-file
                                  %     .fig, .tiff,  .jpg 
                                  %         - one file is created for each
                                  %         figure, appended by its figure
                                  %         index, e.g. 'PhysIO_output_fig01.jpg'
                                 
                                  
    %% ons_secs
    % output structure holding all time-dependent variables, i.e. onsets, specified in seconds
    % all elements but .raw are cropped to the acquisition window of
    % interest
    ons_secs                     = [];
    ons_secs.c              	 = [];  % raw cardiac waveform (ECG or PPU)
    ons_secs.r              	 = [];  % raw respiration volume time course
    ons_secs.t              	 = [];  % time vector corresponding to c and r
    ons_secs.cpulse         	 = [];  % onset times of cardiac pulse events (e.g. R-peaks)
    ons_secs.svolpulse      	 = [];  % [Nscans x 1] onset times of volume scan events
    ons_secs.spulse         	 = [];  % [Nscans*Nslices x 1] onset times of slice (incl. volume) scan events
    ons_secs.spulse_per_vol 	 = [];  % cell(Nscans,1), as spulse, holding slice scan events sorted by volume
    ons_secs.raw            	 = [];  % raw read-in version of the whole structure, before any cropping
end

switch default_scheme
    case 'RETROICOR'
        model.type = 'RETROICOR';
        model.order = struct('c',3,'r',4,'cr',1, 'orthogonalise', 'none');
    case 'redetect_peaks_from_logfile'
        thresh.cardiac.initial_cpulse_select.method = 'manual'; % 'load_from_logfile', 'manual', 'load'
        thresh.cardiac.initial_cpulse_select.file = 'kRpeak.mat';
        thresh.cardiac.initial_cpulse_select.min = 1;
        thresh.cardiac.initial_cpulse_select.kRpeak = [];
    case 'manual_peak_select'
        thresh.cardiac.posthoc_cpulse_select.method = 'manual'; % 'off', 'manual' or 'load',
        thresh.cardiac.posthoc_cpulse_select.file = 'posthoc_cpulse.mat';
        thresh.cardiac.posthoc_cpulse_select.percentile = 80;
        thresh.cardiac.posthoc_cpulse_select.upperThresh = 60;
        thresh.cardiac.posthoc_cpulse_select.lowerThresh = 30;        
end

%% assemble output
physio.log_files   = log_files;
physio.thresh  = thresh;
physio.sqpar   = sqpar;
physio.model   = model;
physio.verbose = verbose;
physio.ons_secs = ons_secs;
